# Table of contents

## 1. Problem Definition & Objectives

This is a transnational dataset that contains all transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based, registered non-store online retail company. The company mainly sells unique all-occasion gifts. Many of the company's customers are wholesalers.

Link to the dataset - [here](https://www.kaggle.com/datasets/carrie1/ecommerce-data).

The primary objective of this project is to develop a user segmentation framework based on the provided dataset, with the aim of **delivering customized offers**. The secondary objective is to conduct an analysis and segmentation for **target campaigns**. The available data includes information such as the types of goods purchased by customers, their geographical locations, and the quantity of goods sold for each type. This data enables detailed customer data segmentation, which is essential for effective marketing strategies.

According to the available data, the upcoming analytical assessment will be <span style="color:red">limited</span> by the following factors:
- Data covers only one year (2010–2011).
- Focus is on UK-based transactions, may not represent global trends.
- Lacks detailed demographic/psychographic information, so segmentation will rely on transaction patterns.

---
**User segmentation** - is the process of separating users into distinct groups, or segments, based on shared characteristics. A company might segment users based on language preferences, product version, geographical region, or user persona.

**The most common types of user segments:**
- **Demographic:** Information about the individual user, such as their age, location, language preferences, title, or role.
- **Firmographic:** Information about the user’s organization, like industry, revenue, employee count, or business model.
- **Technographic:** Information on the other technologies the user’s organization uses, including CRM provider, marketing automation tools, back office systems, or databases.
- **Customer data:** Information stored in a CRM about the customer’s relationship with the company, such as plan type, customer journey stage, annual revenue, account owner, or renewal date.
- **Behavioral:** Information about how the individual user has interacted with the product, including number of logins, pages viewed, features clicked, support tickets created, and time on site.
- **Psychographic:** Information on a user’s likes and dislikes, including product sentiment. This data can be captured with measures like Customer Satisfaction (CSAT) or Net Promoter Score (NPS).

---
The main objective of customer segmentation for the target e-commerce company is to develop unique sales and communication strategies for each segment based on data analysis and customer behavior. Completing the customer segmentation strategy will allow the company to reach its full potential, strengthen consumer loyalty, and optimize the marketing budget. Therefore, the objective of this analysis is to understand the company's heterogeneous customer base. Without segmentation, marketing campaigns are less effective, leading to wasted budget and reduced customer engagement. Thus, the main goals and business results are formulated as follows:
1) Delivering personalized offers - higher conversion rate;
2) Targeted campaigns - reduced customer churn and increased retention.

According to the dataset, **the primary objective** is to **create meaningful customer segments using transaction data**. The secondary goals are to answer the following questions:
- What are the main groups of customers buying from the company?
- Who are the most frequent buyers?
- Who are the highest-spending customers?
- Who are the lower-spending customers?
- Which customers bring in the most profit?

The following deliverables are anticipated:
- Implement a segmentation framework using clustering methods.
- Profiles of customer segments are presented in summary tables and visualizations.
- The final business recommendations are based on the results obtained.


## 2. Data Understanding

Begin the data understanding process by loading the customer dataset, examining the variables, and identifying any missing values.

### 2.1. Load necessary libraries

In [2]:
import numpy as np # linear algebra
import pandas as pd # data processing

### 2.2. Loading the dataset

In [3]:
df = pd.read_csv('D:/Google Drive/Work/Data Analytics and Data Science/Roadpam/project_1-customer-segmentation/data/raw/e-commerce-data-raw.csv', encoding='latin1')
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [4]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,12/9/2011 12:50,4.95,12680.0,France


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Upon loading the dataset, it was ascertained that it determined ***541908*** records (rows) and ***8*** attributes (columns). The columns of the dataset can be described in the following way: 
- 4 columns are a numeric type of data;
- 2 text type (In case of making a dashboard report, it is possible to change the *Country* column to a geographical data type);
- 1 date/time column - *InvoiceDate* which should be transformed into two separate columns *InvoiceDate* and *InvoiceTime*.

The brief description of each column:
|№|Column name|Column description|Data type|Parameters|
|---|---|---|---|---|
|1|InvoiceNo|A unique identifier assigned to each transaction|object|
|2|StockCode|An internationally standardized system of names and numbers to classify traded products|object|
|3|Description|A description of the trade product|object|
|4|Quantity|The quantity of units of a product in a transaction|int64 |
|5|InvoiceDate|The date and time of the transaction|object|
|6|UnitPrice|The unit price of the product is expressed in sterling|float64|
|7|CustomerID|A unique identifier assigned to each customer|float64|
|8|Country|The country of the customer|object|

Before the first step of analysis, 
1. Note to description that CustomerID and Description - have a lot of missing values
2. Transform InvoiceDate to date/time and split
3. Change the alias of the column name

In [10]:
print(f"Initial dtype: {df['InvoiceNo'].dtype}")

Initial dtype: object


In [11]:
df['InvoiceNo'] = pd.to_numeric(df['InvoiceNo'], errors='coerce')
print(f"after dtype: {df['InvoiceNo'].dtype}")

after dtype: float64
